### PR TITLE
onnxruntime: CUDA fixes

### PR DESCRIPTION
**Pull request by autrimpo**
## Description of changes

Fixes for onnxruntime with CUDA from #293768 and helping various people build the package on "normal" hardware.

#293768 adds missing inputs for the python package to successfully auto-patchelf the onnxruntime shared library (thank you @NiklasGollenstede!).

I've moved `gtest` back to its original place - it is only needed for the tests. We disable *running* them for CUDA builds, so the dependency was not being pulled, but it is still required for *building* them. This is now handled correctly - disabling `checkPhase`  also disables building the tests, which should also slightly speed up the build.

Finally, I've set a limit for nvcc threads to 1 - upstream uses a [function in their build script](https://github.com/microsoft/onnxruntime/blob/7ec51f0a13d5d8cbd796de33276bf04210ce6176/tools/ci_build/build.py#L930) to figure out a reasonable value. Given that they expect >= 4GB RAM *per thread*, I don't think trying to tie it to `NIX_BUILD_CORES` is a good idea, as that would eat up all the RAM on most "consumer" machines (8 cores are not that uncommon, 64GB+ RAM on such machines is). In my testing with a couple of builds, I did not notice a significant slowdown with this set, I'd even say it was slightly faster for some reason. The great benefit is many more people should now be able to safely build this.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/libraries/onnxruntime/default.nix [hunk 1]
```diff
@@ -151,6 +150,7 @@ effectiveStdenv.mkDerivation rec {
   ]);
 
   nativeCheckInputs = lib.optionals pythonSupport (with python3Packages; [
+    gtest
```
Comment (jonringer) [inline pkgs/development/libraries/onnxruntime/default.nix:153]: this should probably be moved outside of the pythonSupport flag as it doesn't exist under python package set.

Comment (autrimpo) [reply pkgs/development/libraries/onnxruntime/default.nix:153]: done

File: pkgs/development/libraries/onnxruntime/default.nix
Comment (SomeoneSerge) [inline pkgs/development/libraries/onnxruntime/default.nix:1]: >  I don't think trying to tie it to NIX_BUILD_CORES is a good idea, as that would eat up...


Can't say this is very convincing given that `NIX_BUILD_CORES` can be controlled by the user 🤔

File: pkgs/development/libraries/onnxruntime/default.nix (hunk 1 above)
Comment (SomeoneSerge) [reply pkgs/development/libraries/onnxruntime/default.nix:153]: > I've moved gtest back to its original place - it is only needed for the tests. 

Ugh there it goes again... so, why is it `nativeCheckInputs` and not `checkInputs`, and how does this work if we _hypothetically_ want to enable tests (using `hostPlatform.emulator`, presumably) in cross-compilation?

Comment (autrimpo) [reply pkgs/development/libraries/onnxruntime/default.nix:153]: Was like that before I started touching it 🤷 You're right though, it should probably go in `checkInputs`.

https://github.com/NixOS/nixpkgs/blob/3392575c4008b9a98dc69e68d02a031b379d5a81/pkgs/development/libraries/onnxruntime/default.nix#L129-L130

File: pkgs/development/libraries/onnxruntime/default.nix
Comment (autrimpo) [reply pkgs/development/libraries/onnxruntime/default.nix:1]: But what if the user wants to use multiple cores for the rest of the build? From my testing, nvcc_threads only had effect during parts of the build that were building CUDA objects and not the rest of the C++ compilation. Tying it together severely limits the number of cores you can use unless you have tons of RAM, resulting in net slower builds than using all cores for C++ and 1 for CUDA.

Comment (autrimpo) [reply pkgs/development/libraries/onnxruntime/default.nix:1]: Ideally it would be a separate setting but I'm not aware of anything.

Comment (autrimpo) [reply pkgs/development/libraries/onnxruntime/default.nix:1]: Actually, thinking about it now, in onnx `NIX_BUILD_CORES` is passed as `make -j` in the build. Not sure exactly how the make targets are set up, but wouldn't that hypothetically result in `NIX_BUILD_CORES` CUDA jobs each with `NIX_BUILD_CORES` threads?

File: pkgs/development/libraries/onnxruntime/default.nix (hunk 1 above)
Comment (SomeoneSerge) [reply pkgs/development/libraries/onnxruntime/default.nix:153]: I'm all sleepy, but this is probably relevant and I plan to acquaint tomorrow: https://github.com/NixOS/nixpkgs/pull/206742/files#diff-9538c800780031db3dfa7746f5a36fbbc895c60fdc896385bedd36940927427dR51

File: pkgs/development/libraries/onnxruntime/default.nix
Comment (SomeoneSerge) [reply pkgs/development/libraries/onnxruntime/default.nix:1]: > But what if the user wants to use multiple cores for the rest of the build? From my testing, nvcc_threads only

Oh I see. Sure let's stay with your current solution

**Review by SomeoneSerge — APPROVED** [review]

**Comment by SomeoneSerge** [comment]
```bash
❯ nom build -I nixpkgs=flake:github:autrimpo/nixpkgs/python-onnxruntime-cuda -f '<nixpkgs>' python3Packages.onnxruntime --print-out-paths
Finished at 08:01:38 after 1s
/nix/store/sm9fnczvjzlanydicfm8x106x26w6k12-python3.11-onnxruntime-1.16.3
```

Nixpkgs config:

<details>

```nix
{
  cudaSupport = true;
  cudaCapabilities = [
    "7.0"
    "8.0"
    "8.6"
  ];
  cudaEnableForwardCompat = false;
  allowUnfreePredicate =
    p:
    builtins.all (
      license:
      builtins.elem license.shortName [
        "CUDA EULA"
        "cuDNN EULA"
      ]
    ) (p.meta.licenses or [ p.meta.license ]);
}
```

</details>

**Pull request merged**